packages/ses/src/permits.js

```diff
@@ -1641,8 +1643,8 @@ export const permitted = {
     '@@toStringTag': 'string',
   },
 
-  lockdown: fn,
-  harden: { ...fn, isFake: 'boolean' },
+  lockdown: hermesFn,
+  harden: { ...hermesFn, isFake: 'boolean' },
 
-  '%InitialGetStackString%': fn,
+  '%InitialGetStackString%': hermesFn,
```

Comment:
Why exactly these three?

Reply:
it's these three intrinsics ^ that throw at 

https://github.com/endojs/endo/blob/7a991aa75b06fc418459b5e5d2d08f9fb73ab113/packages/ses/src/intrinsics.js#L101-L104

when calling `lockdown()` on Hermes VM (surfaced in React Native)

Reply:
on Hermes [v0.12.0](https://github.com/facebook/hermes/releases/tag/v0.12.0) it's these three ^
on [v0.13.0 for RN0.75.x](https://github.com/facebook/hermes/releases/tag/v0.13.0) only the next two

otherwise on Static Hermes (unreleased) it's [fixed](https://github.com/facebook/hermes/issues/1371#issuecomment-2050775512)
which atm can only be tested by [building and running](https://github.com/facebook/hermes/blob/main/doc/BuildingAndRunning.md) either [sh_stable](https://github.com/facebook/hermes/tree/sh_stable) or [static_h](https://github.com/facebook/hermes/tree/static_h)

after https://github.com/endojs/endo/pull/2334 lands, adding v0.13.0 to CI is up next (it's [not](https://github.com/facebook/hermes/issues/1371#issuecomment-2341569115) on npm, so will need to wget the .tar.gz)
followed by Static Hermes next (no distro yet so in CI will need to build with cmake)

Reply:
the [initial](https://github.com/endojs/endo/pull/2563/commits/415719fecdafd04e15ea487025dd75d85b1c2485) approach was minimal and to ignore these three, but it doesn't remove them

the proposed change expects `undefined` and sets `undefined` which seems better

or calling `.bind()` on these three instead which also works ([repro](https://github.com/endojs/endo/tree/ses-hermes-completePrototypes-bind))

Reply:
I'm confused. Why doesn't https://github.com/endojs/endo/pull/1221 make this a non-issue, setting the `.prototype` to `undefined` itself?

Reply:
Is it because the test in intrinsics.js is inconsistently stricter than the #1221 test in permit-intrinsics.js and runs first?

I also looked at the definition of these three functions to see what's different there. It looks like both `harden` and `getStackString` are defined as concise methods, so I'm inferring that the hermes bug is that concise methods have a `.prototype`. Does hermes also make these incorrectly constructable (i.e., with `new`)? Is this also fixed in more recent hermeses? (what's the plural of "hermes"?)

But AFAICT `lockdown` is defined as an arrow function. So why is it more problematic than a zillion other arrow functions? Or, perhaps, is it the only intrinsic defined by us as an arrow function?

If all this is true, then it makes sense to permit `prototype: undefined` on these three. So I'll approve on that basis. But before merging, please let us know if this is indeed all true.

Reply:
Separately, it looks like a bug that

> the test in intrinsics.js is inconsistently stricter than the https://github.com/endojs/endo/pull/1221 test in permit-intrinsics.js and runs first

Could you file an issue on that and assign it to me? Thanks.

Reply:
> I also looked at the definition of these three functions to see what's different there.

awesome ^ was about to examine how these three are defined next

> It looks like both harden and getStackString are defined as concise methods, so I'm inferring that the hermes bug is that concise methods have a .prototype.

exactly, but only a bug on Hermes v0.12.0 (and older)

```js
// Hermes release version: 0.12.0, HBC bytecode version: 89
var arrowFn = () => {}
arrowFn.prototype // { [constructor]: [Function] }
```

> Does hermes also make these incorrectly constructable (i.e., with new)?

fortunately not, so only a fake .prototype constructor on concise methods

```js
// Hermes release version: 0.12.0, HBC bytecode version: 89
var arrowFn = () => {}
arrowFn.prototype // { [constructor]: [Function] }
new arrowFn // Uncaught TypeError: Function is not a constructor
```

> Is this also fixed in more recent hermeses? (what's the plural of "hermes"?)

thought this was only fixed in Static Hermes, [mentioned](https://github.com/facebook/hermes/issues/1371#issuecomment-2050775512) in the issue and commit
but from testing it's been back ported to v0.13.0 too!

```js
// Hermes release version: 0.13.0, HBC bytecode version: 96 (bug: fixed)
var arrowFn = () => {}
arrowFn.prototype // undefined
```

```js
// Static Hermes custom version: 25d892b, HBC bytecode version: 98 (bug: fixed)
var arrowFn = () => {}
arrowFn.prototype // undefined
```

Reply:
side note: looks like _all_ Hermes [releases](https://github.com/facebook/hermes/releases) have been published to GitHub now?! (but still not npmjs, if planned)

but it remains there's no need for SES to support these older versions
since metamask-mobile is now on React Native 0.72.15
whereas Hermes v0.11.0 for RN0.68.x has been [unsupported](https://github.com/reactwg/react-native-releases/blob/main/docs/support.md) for years

Reply:
> But AFAICT `lockdown` is defined as an arrow function. So why is it more problematic than a zillion other arrow functions? Or, perhaps, is it the only intrinsic defined by us as an arrow function?

this part's puzzling been puzzling me too...
why is completing the `lockdown` prototype a problem on Hermes v0.12.0 (and older)...
but not on Hermes v0.13.0+ 🤔

> If all this is true, then it makes sense to permit `prototype: undefined` on these three. So I'll approve on that basis. But before merging, please let us know if this is indeed all true.

sure ^ i'll keep digging to find the answer to these last couple questions

Reply:
> I'm confused. Why doesn't #1221 make this a non-issue, setting the `.prototype` to `undefined` itself?

> Is it because the test in intrinsics.js is inconsistently stricter than the https://github.com/endojs/endo/pull/1221 test in permit-intrinsics.js and runs first?

exactly the current order and strictness on `master` is

- intrinsics.js > completePrototypes
  - `isObject(intrinsic)`
  - `objectHasOwnProperty(intrinsic, 'prototype')`
  - `typeof permit === 'object'`
  - `!permit.prototype`
    - ```throw TypeError(`${name}.prototype property not whitelisted`);```
- permit-intrinsics.js > whitelistIntrinsics
  - `prop in obj`
  - `typeof obj === 'function' && prop === 'prototype'`
    - `obj.prototype = undefined` (not reached)
    - #1221

Reply:
> Separately, it looks like a bug that
> 
> > the test in intrinsics.js is inconsistently stricter than the #1221 test in permit-intrinsics.js and runs first
> 
> Could you file an issue on that and assign it to me? Thanks.

- https://github.com/endojs/endo/issues/2598

Reply:
> So why is it more problematic than a zillion other arrow functions?

for example checking: permits.js > Appendix B > Annex B

https://github.com/endojs/endo/blob/4ca4028bb67ba9198060f30a7e5bbff21e8fa27f/packages/ses/src/permits.js#L1613-L1650

these aren't problematic on Hermes
since they're not proposed by SES
so already implemented correctly in Hermes natively:

```js
function escape() { [native code] }
function unescape() { [native code] }
```

however our [three functions proposed by SES](https://github.com/endojs/endo/pull/2563#discussion_r1816606864) are problematic
since we are adding them to Hermes as [buggy](https://github.com/endojs/endo/pull/2563#discussion_r1811111552) arrow functions and concise methods:

```js
function (a0) { [bytecode] } // lockdown
function harden(a0) { [bytecode] } // harden
function getStackString(a0) { [bytecode] } // intrinsic: %InitialGetStackString%
```